plugins/BEdita/Core/src/Shell/StreamsShell.php

```diff
@@ -77,4 +93,83 @@ public function removeOrphans()
+     *
+     * @return void
+     */
+    public function refreshMetadata()
+    {
+        $conditions = [
+            'OR' => [
+                $this->Streams->aliasField('file_size') . ' IS' => null,
+                $this->Streams->aliasField('width') . ' IS' => null,
+                $this->Streams->aliasField('height') . ' IS' => null,
+            ],
+        ];
+        $force = (bool)$this->param('force');
+        if ($force) {
+            $conditions = [];
+        }
+
+        $count = $this->Streams->find('all', compact('conditions'))->count();
+        $this->info(sprintf('Checking %d streams', $count));
+
+        // Set custom error handler because Stream calls some PHP functions that use old error handling
+        set_error_handler(
+            function (int $code, string $message, string $filename, int $lineNumber): void {
+                throw new \ErrorException($message, $code, LOG_ERR, $filename, $lineNumber);
+            }
+        );
+        foreach ($this->streamsGenerator($conditions) as $stream) {
+            $this->updateStreamMetadata($stream);
+        }
+        restore_error_handler();
+    }
+
+    /**
+     * Update stream metadata.
+     *
+     * @param Stream $stream The stream to update
+     * @return void
+     */
+    protected function updateStreamMetadata(Stream $stream): void
+    {
+        try {
+            // Read current file's content...
+            $content = $stream->contents;
+            if ($content === null) {
+                $this->warn(sprintf('  stream %s (object %d) is empty or could not be read', $stream->uuid, $stream->object_id));
+
+                return;
+            }
+
+            // ...and write it back, triggering Stream model's methods to read metadata from file
+            $stream->contents = $content;
+            if ($stream->isDirty() && !$this->Streams->save($stream)) {
+                $this->err(sprintf('  error updating stream %s (object %d): %s', $stream->uuid, $stream->object_id, print_r($stream->getErrors(), true)));
+            }
+        } catch (\Throwable $t) {
+            $this->err(sprintf('  error updating stream %s (object %d): %s', $stream->uuid, $stream->object_id, $t->getMessage()));
+        }
+    }
+
+    /**
+     * Generator to paginate through all streams.
+     *
+     * @param array $conditions Optional filtering conditions
+     * @param int $pageSize Number of objects per page
+     * @return \Generator|Stream[]
+     */
+    protected function streamsGenerator(array $conditions = [], int $pageSize = 100): \Generator
+    {
+        $query = $this->Streams->find('all', compact('conditions'));
+        $pageCount = ceil($query->count() / $pageSize);
+
+        for ($page = 1; $page <= $pageCount; $page++) {
+            yield from $query->page($page, $pageSize)->all();
+            $this->info(sprintf('  processed %d streams', min($page * $pageSize, $query->count())));
+        }
+    }
```

Comment:
This skips a lot of results, because after retrieving page 1 with streams [A B C], the same query does not return [A B C D E F G H I] anymore, but only [D E F G H I]. So, after the first page has been processed, page 2 becomes [G H I], and [D E F] are skipped entirely.

If we refactor this method to accept a `Query` object rather than an array of conditions, we could:

```suggestion
    protected function streamsGenerator(Query $query, int $pageSize = 100): \Generator
    {
        // Although `id` is not a monotonically increasing field, we will at most skip the streams that are created
        // AFTER we launch the script, and whose UUID is lexicographically less than the ID we are currently
        // checking — but we still cover all streams created before our script starts!
        $query = $query->sortAsc($this->Streams->aliasField('id'));
        $q = clone $query;
        do {
            $results = $q->all();
            if ($results->isEmpty()) {
                break;
            }

            yield from $results;

            $last = $results->last();
            $q = clone $query;
            $q = $q->where(function (QueryExpression $exp) use ($last): QueryExpression {
                return $exp->gt($this->Streams->aliasField('id'), $last->id);
            });
        }
    }
```

Reply:
I agree that the approach taken does not work, but we're hitting another problem: `streams` does not have an `id` column, only `uuid` which is a random string.

We could ignore all of it and yield directly from `->all()`:
```php
    protected function streamsGenerator(Query $query, int $pageSize = 100): \Generator
    {
        yield from $query->all();
    }
```

Reply:
I committed the changes, could you look at it again please?